core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java

```diff
@@ -0,0 +1,202 @@
+package org.infinispan.globalstate.impl;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.channels.FileLock;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.infinispan.Cache;
+import org.infinispan.commands.RemoveCacheCommand;
+import org.infinispan.commons.CacheConfigurationException;
+import org.infinispan.commons.api.CacheContainerAdmin;
+import org.infinispan.commons.util.concurrent.ConcurrentHashSet;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.configuration.parsing.ConfigurationBuilderHolder;
+import org.infinispan.configuration.parsing.ParserRegistry;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.PostStart;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.globalstate.GlobalConfigurationManager;
+import org.infinispan.globalstate.ScopeFilter;
+import org.infinispan.globalstate.ScopedState;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.registry.InternalCacheRegistry;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * Implementation of {@link GlobalConfigurationManager}
+ *
+ * @author Tristan Tarrant
+ * @since 9.2
+ */
+
+public class GlobalConfigurationManagerImpl implements GlobalConfigurationManager {
+   private static Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private GlobalConfiguration globalConfiguration;
+   private EmbeddedCacheManager cacheManager;
+   private Cache<ScopedState, Object> stateCache;
+   public static final String CACHE_SCOPE = "cache";
+   private ParserRegistry parserRegistry;
+   private ConcurrentHashSet<String> persistentCaches;
+
+   @Inject
+   public void inject(GlobalConfiguration globalConfiguration, EmbeddedCacheManager cacheManager) {
+      this.globalConfiguration = globalConfiguration;
+      this.cacheManager = cacheManager;
+   }
+
+   @Start
+   public void start() {
+      InternalCacheRegistry internalCacheRegistry = cacheManager.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
+      internalCacheRegistry.registerInternalCache(
+            CONFIG_STATE_CACHE_NAME,
+            new ConfigurationBuilder().build(),
+            EnumSet.of(InternalCacheRegistry.Flag.GLOBAL));
+      parserRegistry = new ParserRegistry();
+      persistentCaches = new ConcurrentHashSet<>();
+   }
+
+   @PostStart
+   public void postStart() {
+      // Initialize caches which are present in the initial state
+      for (Map.Entry<ScopedState, Object> e : getStateCache().entrySet()) {
```

Comment:
The loop here is because I want to ensure that the caches are created before we process the persistent stuff below. With the listener (even with includeCurrentState), this doesn't happen synchronously.